packages/core/src/CorePlugin.ts

```diff
@@ -128,6 +141,121 @@ export class CorePlugin extends JWPlugin {
+        if (range.isCollapsed()) {
+            this.editor.vDocument.insertText(params.range, this.indentChars);
+        } else {
+            const firstCharsOfLines = this.getFirstForEachLineInRange(range);
+            // If there is at least two lines
+            if (firstCharsOfLines.length > 1) {
+                // For each `firstCharsOfLines`, add all the `indentChars` in the document
+                firstCharsOfLines.forEach((firstCharNode: VNode) => {
+                    this.editor.vDocument.withRange(VRange.at(firstCharNode), range => {
+                        this.editor.vDocument.insertText(range, this.indentChars);
+                    });
+                });
+            } else {
+                this.editor.vDocument.insertText(params.range, this.indentChars);
+            }
+        }
+    }
+
+    /**
+     * Outdent lines.
+     *
+     * If there is more than one line selected, for each of the lines, remove up to 4 spaces in the
+     * beggining of the line.
+     */
+    outdent(params: OutdentParams): void {
+        const range = params.range || this.editor.vDocument.selection.range;
+        const firstChars = this.getFirstForEachLineInRange(range);
+        const numberOfSpacePerIndent = this.indentChars.length;
+        // Only outdent when there is at leat two lines selected
+        if (firstChars.length > 1) {
+            firstChars.forEach((firstCharVNode: VNode) => {
+                const spaceChars = [];
+                let currentCharVNode = firstCharVNode;
+                for (let i = 0; i < numberOfSpacePerIndent; i++) {
+                    if (
+                        currentCharVNode &&
+                        isChar(currentCharVNode) &&
+                        /^\s$/g.test(currentCharVNode.name)
+                    ) {
+                        spaceChars.push(currentCharVNode);
+                        currentCharVNode = currentCharVNode.next(
+                            n => n.parent === firstCharVNode.parent && !isMarker(n),
+                        );
+                    } else {
+                        break;
+                    }
+                }
+                spaceChars.forEach(node => node.remove());
+            });
+        }
+    }
+
+    /**
+     * Get the first char for each line in the range.
+     *
+     * Example:
+     * ```html
+     * <p>
+     *  a[bc
+     *  def<br/>
+     *  gh
+     * </p>
+     * <p>
+     *   ij
+     * </p>
+     * <p>
+     *   k]l
+     * </p>
+     * ```
+     * will return the char node: a,d,g,i,k
+     */
+    getFirstForEachLineInRange(range: VRange): VNode[] {
+        const firstChars = [];
+        const lineBreaks = range.selectedNodes.filter(node => node instanceof LineBreakNode);
+        const parents = distinct(
+            range.selectedNodes.filter(n => n.children.length === 0).map(node => node.parent),
+        );
+        if (lineBreaks.length || parents.length) {
```

Comment:
Why testing `lineBreaks.length`? Don't `parents.length` always contain at least one parent?
Also, `lineBreaks.length` serve no purpose because even if `lineBreaks` has no elements, it will simply not add any chars in the instruction below.